src/System.Memory/src/System/Buffers/ReadOnlySequence_helpers.cs

```diff
@@ -226,24 +226,21 @@ private static void CheckEndReachable(object startSegment, object endSegment)
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
         private long GetLength(in SequencePosition start, in SequencePosition end)
         {
-            GetTypeAndIndices(start.GetInteger(), end.GetInteger(), out SequenceType type, out int startIndex, out int endIndex);
-
+            int startIndex = GetIndex(start);
+            int endIndex = GetIndex(end);
             object startObject = start.GetObject();
             object endObject = end.GetObject();
-            if (type == SequenceType.MultiSegment && startObject != endObject)
+            if (startObject != endObject)
             {
                 Debug.Assert(startObject != null);
-                Debug.Assert(startObject is ReadOnlySequenceSegment<T>);
                 Debug.Assert(endObject != null);
-                Debug.Assert(endObject is ReadOnlySequenceSegment<T>);
 
-                var startSegment = Unsafe.As<ReadOnlySequenceSegment<T>>(startObject);
-                var endSegment = Unsafe.As<ReadOnlySequenceSegment<T>>(endObject);
+                var startSegment = (ReadOnlySequenceSegment<T>)startObject;
+                var endSegment = (ReadOnlySequenceSegment<T>)endObject;
                 // (end.RunningIndex + endIndex) - (start.RunningIndex + startIndex) // (End offset) - (start offset)
                 return endSegment.RunningIndex - startSegment.RunningIndex - startIndex + endIndex; // Rearranged to avoid overflow
```

Comment:
@pakrym, can't this overflow/underflow if the `start` or `end` SequencePosition are ill-formed? For example: `int64.MaxValue - 0 - 0 + 1` overflows

If only well-formed SequencePositions are allowed, we wouldn't need to re-arrange the expression to avoid overflow.

Also, is there semantic meaning of a negative length? `0 - 10 + 0 + 0` would return -10.

Reply:
> If only well-formed SequencePositions are allowed, we wouldn't need to re-arrange the expression to avoid overflow.

In theory `end.RunningIndex + endIndex` can overflow even for correct ROS. And we added check for correctness to constructor.

>Also, is there semantic meaning of a negative length? 0 - 10 + 0 + 0 would return -10.
Nope, that's ill-formed ROS

Reply:
So, the GetLength needs to be fixed, correct?

Reply:
> So, the GetLength needs to be fixed, correct?

You can work with endSegment.RunningIndex,  startSegment.RunningIndex, startIndex, endIndex in any order. Result will be the same. For correct structure it will be >= 0. We can just to throw exception if it < 0.

Reply:
> We can just to throw exception if result < 0.

I would be fine with that. @pakrym, thoughts?
```C#
int endOffset = (end.RunningIndex + endIndex);
int startOffset = (start.RunningIndex + startIndex);
if (endOffset < 0 || startOffset < 0) throw;
return endOffset - startOffset;
```

Reply:
`int endOffset = (end.RunningIndex + endIndex);` should be long.
Also I suggest not add checking for RunningIndex because all will be work correct without such checking. And we will have more faster code.

Reply:
> Are you going to add a test for this scenario? :)

Thinking about it some more, the current implementation is fine.
In fact, writing it as `return (end.RunningIndex + endIndex) - (start.RunningIndex + startIndex) `should also be fine.

We already guarantee that `end.RunningIndex >= start.RunningIndex` in the ctor and hence `(end.RunningIndex + endIndex) >= (start.RunningIndex + startIndex)` is also guaranteed (especially since `startIndex <= start.RunningIndex` and `endIndex <= end.RunningIndex`). The only time we would have negative length is if someone had a buffer of size larger than `(long.MaxValue - int.MaxValue)` which isn't practical.

Reply:
I tested these cases to confirm:
```C#
var bufferSegmentA = new BufferSegment<T>(new T[99]);
BufferSegment<T> bufferSegmentB = bufferSegmentA.Append(new T[100]);
BufferSegment<T> bufferSegmentC = bufferSegmentB.Append(new T[50]);

for (int i = 0; i <= bufferSegmentA.Memory.Length; i++)
{
    for (int j = 0; j <= bufferSegmentB.Memory.Length; j++)
    {
        var myBuffer = new ReadOnlySequence<T>(bufferSegmentA, i, bufferSegmentB, j);
        Assert.True(myBuffer.Length >= 0);
    }
}

for (int i = 0; i <= bufferSegmentA.Memory.Length; i++)
{
    for (int j = 0; j <= bufferSegmentC.Memory.Length; j++)
    {
        var myBuffer = new ReadOnlySequence<T>(bufferSegmentA, i, bufferSegmentC, j);
        Assert.True(myBuffer.Length >= 0);
    }
}

for (int i = 0; i <= bufferSegmentB.Memory.Length; i++)
{
    for (int j = 0; j <= bufferSegmentC.Memory.Length; j++)
    {
        var myBuffer = new ReadOnlySequence<T>(bufferSegmentB, i, bufferSegmentC, j);
        Assert.True(myBuffer.Length >= 0);
    }
}
```